### ISSUE TITLE
Parser Annotation Errors megaticket

### ISSUE THREAD

**Issue by ChrisPenner** [unisonweb/unison#3639]
There are several subtle parser bugs which end up with ABT nodes that either have incorrect annotations, or have an annotation which does not correctly span the ranges of the nodes they contain.

This has been largely ignored, since many of these spans are "close enough" for error messages, or because the spans simply weren't used for anything until now, but I need them to be correct for the LSP to map a source file range to an ABT node.

They also need to be correct for things like auto-formatting to work correctly in LSP.

1. `effect` and `effect1` in `Unison.Type` erroneously tag their effects list with the annotation for the remainder of the arrow, this results in false matches when hovering on type signatures

E.g. Hovering over `Nat` in the following will match a hidden and empty effects list which contains the right -side of the arrow.

```unison
thingy : Text -> Nat
thingy  = do 1 + 1
```

2. Top level bindings with args left of the `=` seem to shadow the top-level type annotation underneath a lambda somehow, e.g. hover over 'Nat' here doesn't work, since it somehow matches the lambda for `User _ -> ...`;
```unison
thingy :  Text -> Nat
thingy _ = 1
```

3. Annotations aren't combined correctly in do-blocks/let-blocks

* [x] https://github.com/unisonweb/unison/pull/3689

E.g. you can't hover anything other than the first line:

```unison
thingy : Nat
thingy = let
  1 + 1
  23
  4556
```

4. Binary expressions don't contain the annotation for their left-hand-side; 
e.g. you can hover over `false` but not `true`

- [x] https://github.com/unisonweb/unison/pull/3640

```unison
thingy : Boolean
thingy = true && false
```

5. Annotations on destructuring binds

Annotations aren't right in expressions like this:

```unison
term = let
    (first, second) = (false, true)
    (third, fourth) = (false, true)
    first && second && third && fourth
```

6. The typechecker adds type annotations with Types that have in-file annotations even though the types aren't defined in the file (they're just inferred). Can probably change these to be some sort of "Derived" type ann, but want to make sure this doesn't negatively effect any error messages.

**Comment by aryairani** [comment]
In progress (@ChrisPenner)

**Comment by ChrisPenner** [comment]
The remaining issues (that I know of at the moment) are resolved in #3795 and #3789 , so I'll close this 😄
